### ISSUE TITLE
Style guide: Thoughts on Enums

### ISSUE THREAD

**Issue by lukesneeringer** [bufbuild/buf#6]
I wanted to share a few thoughts on enums that may justify some minor tweaks to the style guide.

First of all, in your section on using `FOO_BAR_UNSPECIFIED` as the first value, the "Why?" page omits something extremely important, which is "why use `FOO_BAR_UNSPECIFIED` rather than just `UNSPECIFIED`"?

The reason: The C++ proto implementation causes all enum values to share the namespace of the enum, so...

```proto
enum Foo {
  UNSPECIFIED = 0;
}
enum Bar {
  UNSPECIFIED = 0; // Boom! These conflict.
}
```

This is unique to C++; it works fine in every other language.

This leads to one point in your style guide that I quibble with, which is the recommendation to avoid nesting enums. Because of this namespace issue, I have actually started recommending to my teams that they _do_ nest enums, particularly State enums, when they can have reasonable confidence. In other words, I would much rather see...

```proto
message Foo {
  enum State {
    STATE_UNSPECIFIED = 0;
    ACTIVE = 1;
    RESTING = 2;
  }
  State state = 99;
}
```

...than the recommendation in the buf guide:

```proto
enum FooState {
  FOO_STATE_UNSPECIFIED = 0;
  ACTIVE = 1;
  RESTING = 2;
}
message Foo {
  FooState foo_state = 99;
}
```

Basically, the risk of accidental namespace collision feels (to me) greater than the concern that you might have to use a nested enum outside of its message later, particularly since `Foo.State` is okay in protobuf.

If nested enums are a no-no, then I would recommend prefixing _every_ value with the enum name, not just the unspecified value:

```proto
enum FooState {
  FOO_STATE_UNSPECIFIED = 0;
  FOO_STATE_ACTIVE = 1;
  FOO_STATE_RESTING = 2;
}
```

However, then you are forcing that effect on users, which I feel is terrible. (Generally users are never typing the unspecified value), so on balance I think nesting enums is usually preferable provided that the enum really is a property of the enclosing message.

Just my two cents; up to your folks what your own style guide is. :-)

**Comment by bufdev** [comment]
> This is unique to C++; it works fine in every other language.

This isn't a C++ problem - as we talk about in the Buf docs, the issue will lead to protoc not compiling definitions (regardless of whether it was meant to only protect against C++ issues). Your example above won't compile:

```
$ protoc -o /dev/null foo.proto
foo.proto:9:3: "UNSPECIFIED" is already defined in "foo".
foo.proto:9:3: Note that enum values use C++ scoping rules, meaning that enum values are siblings of their type, not children of it.  Therefore, "UNSPECIFIED" must be unique within "foo", not just within "Bar".
```

We also find that nesting enums leads to issues with forwards compatibility that while not actually breaking compatibility, breaks the spirit, namely that developers typically think an enum only truly applies to their specific message, only later realizing it is a more global concept. This leads to references in other messages for nested enums, which is worse than standardizing on a prefix.

The Google Style Guide also uses it for their example https://developers.google.com/protocol-buffers/docs/style

We'll stick with this recommendation - we find that many people have opinions on Protobuf style, but the one that leads to the least amount of issues down the road, even if more verbose, is usually preferred, especially to help developers of all levels of familiarity with Protobuf.

**Comment by lukesneeringer** [comment]
That sounds reasonable. :-)

**Comment by noahseger** [comment]
@bufdev your comment about nesting is valuable and worth referencing, because the prefix lint rule is the most hated on our team :D Would it make sense to add to https://buf.build/docs/best-practices/style-guide#enums?
